CI enabled flag

kristapratico opened the pull request:
Addresses part of https://github.com/Azure/azure-sdk-for-python/issues/33256

This adds a `ci_enabled` flag to the pyproject.toml configuration. It defaults to `True`. If a service sets `ci_enabled = false`, analyze checks and tests will be disabled when running a Manual or internal build. Pull request builds will continue to run all checks in order to prevent code being checked in on a failing pipeline. Special case: release builds will fail if `ci_enabled=false` to prevent a package release without checks running.

**Motivation**: As we work towards enabling some checks like mypy, strict_sphinx, etc. it is evident that not all libraries will be compliant at deadlines. We still want to turn these checks to mandatory for all libraries, but not at the expense of breaking a ton of library pipelines and impacting pipeline health. This pyproject.toml option will allow us to disable failing pipelines, but still perform validation at Pull Request time to ensure failing checks get addressed before code changes are made. It also makes more clear/trackable which pipelines are disabled (rather than disabling under the DevOps pipeline settings) and easy to turn back on when an owner is ready to address issues.

**Pending work / Not addressed in this PR**: Right now we disable analyze checks and tests, but the skip is done *after* we allocate resources to create the jobs. To fully implement this, when the config is set, we shouldn't waste resources and just skip spinning up the entire job.

**Test runs** (using metricsadvisor as an example where we do want to skip checks):

- PR `- ci` run: https://dev.azure.com/azure-sdk/public/_build/results?buildId=3326875&view=results
    - Expected: All checks run
- Manual `- ci` run: https://dev.azure.com/azure-sdk/public/_build/results?buildId=3327018&view=results
    - Expected: Analyze checks and tests skipped
- Manual or scheduled internal (release) run: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3327024&view=results
    - Expected: Analyze checks and tests skipped
- Manual or scheduled internal (`- tests`) run: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3327023&view=results
    - Expected: Analyze checks and tests skipped

**Test release**

- ci_enabled = true: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3333464&view=results
    - Expected: Successful release
- ci_enabled = false: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3333587&view=logs&j=7bb69cc7-6f52-585e-8b13-5bae7c4a9fa4&t=66ec37bc-d774-526e-7316-aa88863bcdc6
    - Expected: Block on release in the `Verify CI enabled` step

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[azure-ai-textanalytics](https://apiview.dev/Assemblies/Review/625c817292d94003afffce4f2e8aa52d)

scbedd commented on eng/pipelines/templates/steps/build-extended-artifacts.yml:
```diff
@@ -86,10 +86,14 @@ steps:
 
+  - pwsh: |
+      $directoryExists = Test-Path -Path "$(Build.SourcesDirectory)/_docs"
+      Write-Output "##vso[task.setvariable variable=DirectoryExists]$directoryExists"
```
Add a `displayName: ` so the step isn't just `powershell` in the task list.

scbedd commented on eng/pipelines/templates/steps/targeting-string-resolve.yml:
```diff
@@ -14,5 +14,12 @@ steps:
       Write-Host "##vso[task.setvariable variable=TargetingString]$setting"
+
+      Write-Host "Build reason: $(Build.Reason)"
+      if ("$(Build.Reason)" -eq 'PullRequest') {
```
Let's remove this. Handle the build-system-specific values of `Build.Reason` in the function `public_ci()`.

If you look at `in_ci()` what we do is return a specific non-zero number for each system. So we _can_ just check non-falsey for `are we in CI`, but also have the detail if we need it.

Please do the same in `in_public()`.

you should be able to check for the presence of `BUILD_REASON` for azure devops, and `GITHUB_EVENT_NAME ` in gh-actions.

according to chatgpt
```
As of my last knowledge update in January 2022, the possible values for GITHUB_EVENT_NAME in GitHub Actions include:

push: Triggered on push events, such as when commits are pushed to a repository branch.
pull_request: Triggered when a pull request is opened, synchronized, or closed.
release: Triggered when a release is created or edited.
schedule: Triggered on a schedule defined in the workflow configuration.
workflow_run: Triggered when a workflow run is requested or completed.
repository_dispatch: Triggered when a repository_dispatch event is received.
```

You just need to recognize `pull_request`.

Doing the above will get rid of the arbitrary variable set + arbitrary variable check -> direct to what the system actually uses.

kristapratico replied:
Done, this is a much better idea

scbedd reviewed (changes requested):
Thanks @kristapratico !

This is most of the way there, but I've added some feedback on variable usage + I sent along an offline question about releasing disabled packages.

kristapratico commented:
> Thanks @kristapratico !
> 
> This is most of the way there, but I've added some feedback on variable usage + I sent along an offline question about releasing disabled packages.

Thanks @scbedd! I think I've addressed your PR + offline feedback so this should be ready for re-review.

scbedd commented on eng/pipelines/templates/stages/archetype-python-release.yml:
```diff
@@ -46,6 +46,14 @@ stages:
                           PackageName: ${{artifact.name}}
                           ServiceName: ${{parameters.ServiceDirectory}}
                           ForRelease: true
+
```
I could be wrong, but I don't think this will actually work without installing `tools/azure-sdk-tools` Have you tried an azure-template release using this?

kristapratico replied:
@scbedd sorry just seeing this comment. I think we discussed this offline. Did your debugging make any sense of it?

scbedd replied:
Yeah...I have no explanation.

scbedd replied:
[Yep. It's there](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3343636&view=logs&j=7bb69cc7-6f52-585e-8b13-5bae7c4a9fa4&t=7abbbe6b-d005-5c95-feb6-99ffe4e699e1&l=140)

scbedd reviewed (approved):
Thanks @kristapratico !
